synapse/storage/databases/main/relations.py

```diff
@@ -978,6 +984,59 @@ def _get_thread_id(txn: LoggingTransaction) -> str:
 
         return await self.db_pool.runInteraction("get_thread_id", _get_thread_id)
 
+    @cached()
+    async def get_thread_id_for_receipts(self, event_id: str) -> str:
+        """
+        Get the thread ID for an event by traversing to the top-most related event
+        and confirming any children events form a thread.
+
+        Given the following DAG:
+
+            A <---[m.thread]-- B <--[m.annotation]-- C
+            ^
+            |--[m.reference]-- D <--[m.annotation]-- E
+
+        It considers A, B, C, D, and E as part of the thread.
+
+        See also get_thread_id.
+
+        Args:
+            event_id: The event ID to fetch the thread ID for.
+
+        Returns:
+            The event ID of the root event in the thread, if this event is part
+            of a thread. "main", otherwise.
+        """
+
+        # Recurse up to the *root* node, then select relations of that to
+        # see if there are thread children.
+        sql = """
+        SELECT relates_to_id FROM event_relations WHERE relates_to_id = COALESCE((
+            WITH RECURSIVE related_events AS (
+                SELECT event_id, relates_to_id, relation_type, 0 depth
+                FROM event_relations
+                WHERE event_id = ?
+                UNION SELECT e.event_id, e.relates_to_id, e.relation_type, depth + 1
+                FROM event_relations e
+                INNER JOIN related_events r ON r.relates_to_id = e.event_id
+                WHERE depth <= 3
+            ) SELECT relates_to_id FROM related_events ORDER BY depth DESC LIMIT 1
+        ), ?) AND relation_type = 'm.thread' LIMIT 1;
+        """
```

Comment:
1. I'm happy that the WITH RECURSIVE expression recurses 3 steps up the relations tree.
2. We then SELECT the `relates_to_id` of the last thing we saw in that WITH RECURSIVE expression.
3. We then look for a relation of type m.thread pointing at the event from 2.

Does this work if event A relates to some even older event X? E.g. suppose that A was sent in reply to X. Wouldn't that mean that in point 2 above we select the event id of X?

Reply:
I'm not really sure what you're asking. I think you're asking is what would happen if you have:

```mermaid
flowchart RL
    A-.->|m.reference|X
    B-.->|m.thread|A
```

This is an invalid structure -- a thread event cannot apply to an event which has a relation itself (i.e. the relation between `B` and `A` is invalid).

Reply:
(I should probably include a bit more of a definition about how this query works though...)

Reply:
> I think you're asking is what would happen if you have:

Yes, exactly this.

> This is an invalid structure 

Oh, well that makes life much easier! And I guess it means we can say that threads are closed under relations, which seems like a neat invariant.